### ISSUE TITLE
Any plan for training code?

### ISSUE THREAD

**Issue by shileims** [lucidrains/perfusion-pytorch#5]
Hi Author,
    This is a really amazing repo.
    I check the progress every day and hope to get it to try.
    Do you have an estimation deadline for finishing the training code?
    Thank you so much!

**Comment by lucidrains** [comment]
hey yea, i'll circle back to it

you should give it a try though. hard part is done and all the pieces are there, for both single and multi-concept

**Comment by IndolentKheper** [comment]
Would you mind giving a little advice on how to run it for us non-developers? Like I understand how to install and all that, but actually running the training isn't going to be easy for me and a lot of others to figure out on our own.

**Comment by ViddleShtix** [comment]
+1 as I am also very interested if you have the time. I'm mostly wondering the broad order of what has to be done, and how we can go about saving the "100kb file" and apply it to a diffusers pipeline.

**Comment by lucidrains** [comment]
i would recommend starting with xiao's dreambooth / SD repository first, since that is what the authors used. Yoad also offered to further debug once we had something wired together

at this point, you'll still need to find a developer / researcher / ML engineer. i'll try to get it to the point where it is just a command line when i find time

**Comment by lucidrains** [comment]
broad order, for any entry level ML engineer who is looking to up their skills

~1. run a subset of laion dataset texts through [the function for generating the input covariance](https://github.com/lucidrains/perfusion-pytorch/blob/main/perfusion_pytorch/perfusion.py#L32)~
~2. determine the super concept id (must be 1 token id) from embedding table~
3. wrap the open `CLIP` instance with `OpenClipEmbedWrapper`, passing it also your `superclass_string`
~4. modify the BPE string to id mapping for your new \*concept\*~
5. wrap all keys and values in the cross attentions with the `Rank1EditModule`
6. pass the outputs from the `OpenClipEmbedWrapper` through the cross attention, and into the `Rank1EditModule` from step 5
7. use `get_finetune_optimizer` on your wired up stable diffusion instance. it should automatically extract only the small number of finetuneable parameters
8. the 100kb of data per fine tuned concepts can be extracted and restored with [`save` and `load` functions](https://github.com/lucidrains/perfusion-pytorch/blob/main/perfusion_pytorch/save_load.py) (just pass it in the wired stable diffusion instance)

**Comment by lucidrains** [comment]
i realize there's still a lot of steps, i'll try to do a bit more damage next week. just writing that up, i realize for number 6, i could also take care of the whole fetching the embedding with the concept substituted with super class concept in the embedding wrapper.

**Comment by lucidrains** [comment]
actually, i'll take care of 6 tomorrow morning. that's probably the remaining tool needed before operating on SD

**Comment by lucidrains** [comment]
the input covariance also shouldn't change, so perhaps someone can run this for the clip being used in SD, and we can just check in the tensor, use it as a default. that would save us from passing around this extra matrix

**Comment by ViddleShtix** [comment]
Here's what I got for the input covariance by taking the first 100k captions from the LAION 400m dataset and running it through the function (perhaps it can be verified in the future):
`tensor([[ 9.3166e+00, -4.6043e-02, -1.1625e-01,  ...,  6.4781e-02,
         -1.4543e+00,  8.6031e-01],
        [-4.6043e-02,  1.0780e+01, -2.8464e-01,  ...,  7.0342e-01,
         -2.0617e-01, -3.6634e-01],
        [-1.1625e-01, -2.8464e-01,  1.0130e+01,  ...,  3.4655e-01,
         -2.7999e-01,  2.1443e-03],
        ...,
        [ 6.4781e-02,  7.0342e-01,  3.4655e-01,  ...,  1.0315e+01,
         -6.7360e-01,  1.0505e+00],
        [-1.4543e+00, -2.0617e-01, -2.7999e-01,  ..., -6.7360e-01,
          1.2800e+01, -2.2284e+00],
        [ 8.6031e-01, -3.6634e-01,  2.1443e-03,  ...,  1.0505e+00,
         -2.2284e+00,  1.4173e+01]])`

Please note there is a slight bug in the calculate function. The first line needs to be removed since it tries to take in all the data instead of batching it, causing it to max out memory and crash.

I calculated the covariance on the ViT-L-14 CLIP model since that's what SD 1.5 uses. My OpenClipAdapter looks like this:
`OpenClipAdapter(name="ViT-L-14", pretrained="laion400m_e32", tokenizer_name="ViT-L-14")`

I saved the output with torch.save and have attached it here. Hope this helps.

[input_covariance_ViT-L-14.zip](https://github.com/lucidrains/perfusion-pytorch/files/12435696/input_covariance_ViT-L-14.zip)

**Comment by lucidrains** [comment]
@BradVidler amazing Brad! thank you so much 🙏 

i'll get this checked in

**Comment by lucidrains** [comment]
@BradVidler checked it in for version 0.1.6! we can now scratch the first step off the list haha

**Comment by lucidrains** [comment]
also addressed returning the prompt with superclass embed, if in training mode

**Comment by lucidrains** [comment]
also added a `save` and `load` function this morning. if you read the file, should be self explanatory. since the keys are locked to the super class concept output, there is still room to whittle down the size of the saved package even more

will get around to it next week

**Comment by irowberry** [comment]
Would love to contribute where I can, it looks like you're knocking out that list quickly though.

**Comment by lucidrains** [comment]
@irowberry sure, you can submit a PR at any time, or try wiring it up to SD 1.5 and give it a go

**Comment by ViddleShtix** [comment]
I believe step 2 can be achieved with the diffusers library like so:
from diffusers import StableDiffusionPipeline
pipeline = StableDiffusionPipeline.from_pretrained(f"runwayml/stable-diffusion-v1-5")
pipeline.to("cuda")
super_concept = "person"
super_concept_token = pipeline.tokenizer.encode(super_concept)[1:-1]
assert len(super_concept_token) == 1, "Super concept token should be 1 token."
print(super_concept_token)

I'm using a slice because the first and last tokens are automatically start/end tokens, like this:
[49406, 2533, 49407]

If it makes sense, open clip can be used as well with the tokenize function:
from open_clip import tokenizer
tokenizer.tokenize("person")

I assumed that going straight to the tokenizer that is loaded with the diffusers model is the better option to ensure we are working with the correct tokenizer version, but both ways end up with the same tokens.

**Comment by lucidrains** [comment]
@BradVidler hey Brad! yes, that is correct, and I can take care of that, since a number of text-to-image models all use the same open clip tokenizer. we can just use the one from `open_clip`

it occurred to me (thanks to your comment) that we can strike off both 2 and 4, by simply having all the prompts include only the superclass concept string. so, if i were trying to perfuse my dog into the weights, i would simply prepare a bunch of prompts with the superclass string `dog`, and at `forward`, substitute `dog` with the new concept id. Therefore, no modification of the BPE dictionary is needed

the `EmbeddingWrapper` would simply receive the `superclass_concept_string: str` as well as the `tokenizer_encode: callable`, and when fine tuning, you would just use `dog`

edit: probably a bit confusing, but let me just get the code down and it will be self explanatory

**Comment by lucidrains** [comment]
https://github.com/lucidrains/perfusion-pytorch/commit/03059d94b0ed3be1fe1d4ce5262ca99b287acdba this should knock off both 2 and 4 above

welcome any code reviews

**Comment by lucidrains** [comment]
i'm also going to cross off 8, as you just use the `save` and `load` function

**Comment by lucidrains** [comment]
ok, will get back to this next week. have a happy weekend!

**Comment by shileims** [comment]
Hi @lucidrains ,
     Would you have a plan to complete the training based on SD1.5? If possible, would you try to add some instructions here to finish the code? I get some days' rest, so I will try to help here. Thanks

**Comment by irowberry** [comment]
@shileims I've almost finished some VERY rough training code. The last step that I'm working on is finding a way to take the output embeddings from the `EmbeddingWrapper` class and pass them through the rest of the text encoder to get the final text encodings that are then passed into the `Rank1EditModules`.

**Comment by shileims** [comment]
Hi @irowberry ,
    You are really awesome!

**Comment by lucidrains** [comment]
always the github anon (rather than that github account that flexes a bunch of titles) that delivers

**Comment by irowberry** [comment]
Just completed a successful forward and backward pass, still need to find out a way to pass in the modified embeddings to the rest of the text encoder model, but hopefully I can train a full test model soon. I counted the number of trainable params and compared it to the total amount, it's about 14% of the total weights (142243008 vs 1001776452) not sure if this is correct?

**Comment by shileims** [comment]
hi @irowberry , the paper justifies that the extra weights are 100kb. Is it useful here?

**Comment by lucidrains** [comment]
@irowberry nice! yea, i think weaving the text token ids through the clip text encoder is tricky, as you'll need to do some model surgery to omit the wrapped embedding. maybe i should offer some function where you pass in the path to the text encoder embedding, and it returns the wrapped version, and also substitutes the original one with a `nn.Identity()`

so 14% does not sound right. the number of parameters being optimized should be miniscule. are you using [this function](https://github.com/lucidrains/perfusion-pytorch/blob/main/perfusion_pytorch/save_load.py#L20) to fetch the parameters for the optimizer?

**Comment by irowberry** [comment]
@lucidrains I am using that function but I did forget to freeze the text encoder layers (whoops), that 14% sounded too big to me too. Now it's just 1.9% of the total parameters (19182528 / 1001776452).

**Comment by lucidrains** [comment]
@irowberry 🚀 bring it home

**Comment by lucidrains** [comment]
@irowberry added [yet another wrapper](https://github.com/lucidrains/perfusion-pytorch/commit/be4f634f31a0bdbaa86cb09b8a9478be9e1eb98b) that should help

**Comment by shileims** [comment]
@irowberry  @lucidrains ,
By using the function "save" in save_load.py file.
I get 50304 trainable parameters over 1066261035 parameters. 
The .pth file has 210KB and 2514KB without and with C_inv respectively.
Any suggestions here? The paper says 100KB...........

**Comment by lucidrains** [comment]
@shileims close enough! I believe it is because I am redundantly saving the concept outputs for the keys across all cross attention layers. since it is locked, it only needs to be stored once

I can work on whittling down the size once / if you see some initial results

**Comment by shileims** [comment]
Hi @lucidrains , thanks for your reply. I will try to hookup the pipeline. Thanks

**Comment by irowberry** [comment]
Thoughts on loss functions? I'm using MSE to compare the noise to the predicted noise, `loss = F.mse_loss(pred, target).mean()` However, I've seen a lot of models use the mask as well, I'm not sure which one to use for this.

**Comment by shileims** [comment]
Follow question: how to use text_mask, attention_mask and image_mask in the pipeline? I am a little bit confused.

**Comment by lucidrains** [comment]
@irowberry just stick with the normal loss for now until you see some signal

we can see how good the technique is on its own

**Comment by lucidrains** [comment]
@shileims yea, masking is the most confusing part of learning attention

there is only one mask you need to worry about. the `text_mask` becomes the `key_padding_mask` in the cross attention layer (as the text tokens itself becomes the key / values). in other words, `text_mask` and the `attention_mask` is the same

the `image_mask` is something different. it is a segmentation mask outputted by another pretrained model, meant to help better guide the model during fine tuning by modulating the loss. basically what @irowberry asked about in the comment above. i say just ignore that for now

**Comment by shileims** [comment]
hi @lucidrains ,
    But in the MemoryEfficientCrossAttention class from official SD repo, it shows the mask should be None.

![image](https://github.com/lucidrains/perfusion-pytorch/assets/104052591/d137b49c-d9e7-42af-9000-8ea442d6338e)

**Comment by lucidrains** [comment]
@shileims oh that's interesting

how about the regular cross attention class?

**Comment by shileims** [comment]
Hi @lucidrains yes, the regular cross attention class copes with the mask. 
![image](https://github.com/lucidrains/perfusion-pytorch/assets/104052591/fe1d22c7-8c23-4815-8946-0e480bdfccef)


Should we switch to regular cross attention?

**Comment by lucidrains** [comment]
@shileims yes, i think so

i'm actually surprised the memory efficient cross attention would work without masking, unless if it is intended for inference with one batch sample at a time (no key padding)

**Comment by shileims** [comment]
hi @lucidrains ,
    Does ema_concept_text_encs require gradients? I think it is i*.
    self.register_buffer('ema_concept_text_encs', torch.zeros(num_concepts, dim_input, **requires_grad=True**))

**Comment by lucidrains** [comment]
@shileims no it does not, the authors update that using exponential moving average afaict

**Comment by shileims** [comment]
hi @lucidrains , @irowberry ,
     Do you have a repo recommendation for integrating current rank-one model editing functions for the dream booth?

**Comment by shileims** [comment]
After implementation by myself, I found the same prompt with different runs will get totally the same results. Any suggestions here? Thanks

**Comment by irowberry** [comment]
Okay, I am not sure what I'm doing wrong, but these are the sorts of images I'm getting. 
![image3](https://github.com/lucidrains/perfusion-pytorch/assets/61517961/329a84dc-f6b5-4833-929d-96b1a9a3bae7)
![image0](https://github.com/lucidrains/perfusion-pytorch/assets/61517961/2571a88e-1395-466f-989a-0861be417f47)
![image1-2](https://github.com/lucidrains/perfusion-pytorch/assets/61517961/c9a42b50-e7a1-4da3-b66d-da5809abc596)

**Comment by lucidrains** [comment]
@irowberry nice! want to share your code? you and @shileims should also get a discord room and review each other's work

**Comment by lucidrains** [comment]
i'm in the middle of a project and can't context switch atm, but will try to get back to this this weekend

**Comment by irowberry** [comment]
Here's the code, I'm trying to get it to work with Hugging Face's models, and it does sort of. However, there's an important edit that needs to be made within the diffusers package. In `diffusers/models/attention_processor.py` in the `Attention` class, and `prepare_attention_mask` method, there is a TODO that says something about stable-diffusion-pipelines (line 408). You just need to change attention_mask to be  `F.pad(attention_mask, (0, target_length - current_length), value=0.0)` in that if branch. 
[train_perfusion.txt](https://github.com/lucidrains/perfusion-pytorch/files/12552849/train_perfusion.txt)

**Comment by lucidrains** [comment]
@irowberry cool, want to check in that text file into either a gist or repo?

**Comment by irowberry** [comment]
Yeah I just made a repo. Here's the link https://github.com/irowberry/perfusion-training/blob/main/train_perfusion.py

**Comment by shileims** [comment]
![image](https://github.com/lucidrains/perfusion-pytorch/assets/104052591/6f84a9e9-6046-4754-bef0-986297fd5d19)

Sometimes, I got image result like this...........

@lucidrains 
Just kindly ask, any chance to get a code review from you?

**Comment by lucidrains** [comment]
@shileims haha! what concepts did you finetune?

yea, happy to code review this weekend, please do the same as @irowberry and share your code online

**Comment by shileims** [comment]
Hi @lucidrains ,
     Really appreciate your code review, that will be very helpful.
     This is the repo link:  https://github.com/shileims/Finetuning/tree/main, main.py is the entrance file. some results are in ckpts folder.  The training concept is "teddy".............

**Comment by yoshibenjo** [comment]
as a side note, i'm more than happy to donate compute to train this model once this review is ready.

**Comment by ViddleShtix** [comment]
Any update on this?

**Comment by ViddleShtix** [comment]
@irowberry I have been testing your code. I think there's something weird happening in the inference. I'm using the following code to test before and after, using a seed.

```
from perfusion_pytorch.embedding import OpenClipEmbedWrapper, EmbeddingWrapper
from perfusion_pytorch.save_load import load
from perfusion_pytorch import Rank1EditModule
import torch
from diffusers import StableDiffusionPipeline

device = "cuda:0"
generator = torch.Generator(device="cuda").manual_seed(12345)
prompts = ["photo of a person, clear face, high quality"]
superclass_string = 'person'

pipe = StableDiffusionPipeline.from_pretrained("runwayml/stable-diffusion-v1-5", requires_safety_checker=False)
pipe.to(device)
image = pipe(prompts[0], num_inference_steps=30, guidance_scale=7.5,generator=generator).images[0]
image.save("test-before.png")

pipe.to("cpu")
perfusion_model = PerfusionModel(pipe.unet, pipe.text_encoder, pipe.tokenizer, superclass_string).to(device).requires_grad_(False)
load(perfusion_model, 'me_concept.pt')
perfusion_model.eval()

pipe.text_encoder.to("cpu")
wrapped_clip_with_new_concept = OpenClipEmbedWrapper(
    pipe.text_encoder,
    tokenizer = pipe.tokenizer,
    superclass_string = superclass_string
)

text_enc, superclass_enc, mask, indices = wrapped_clip_with_new_concept(prompts)

pipe = StableDiffusionPipeline.from_pretrained("runwayml/stable-diffusion-v1-5", requires_safety_checker=False)
pipe.unet = perfusion_model.unet
pipe.text_encoder = perfusion_model.clip_model
pipe.to(device)

image = pipe(prompts[0], num_inference_steps=30, guidance_scale=7.5,generator=generator,cross_attention_kwargs={"concept_indices": indices, "text_enc_with_superclass": superclass_enc}).images[0]
image.save("test-after.png")
```

Unfortunately the results don't look like the subject (me) so I think I am missing something, or something is going wrong during training. 

I tested @shileims code briefly as well and it also doesn't seem to be training the subject into the model unless I am missing something there as well.

One theory: It seems like we are using the same concept and superclass string, eg. person, so when we train on "person", we don't end up seeing much of a change because that token is so common. If for example we could train on bradvidler and then and link that to the superclass person, that would work better since bradvidler is a rare token. I think in the original Dreambooth paper they suggest using a rare token, 3-4 characters, eg. ohwx. In the perfusion paper they use eg. Hugsy -> Teddy.

If @lucidrains is able to comment on this that would be much appreciated. I am specifically looking for clarity on the earlier comment here https://github.com/lucidrains/perfusion-pytorch/issues/5#issuecomment-1694393325. Thanks in advance!

**Comment by irowberry** [comment]
@BradVidler are you getting somewhat coherent images out? Mine are always abstract shapes. I followed code straight from Hugging Face for inference, which is found [here](https://huggingface.co/docs/diffusers/main/en/using-diffusers/write_own_pipeline)  I'll keep working on it.

> Okay, I am not sure what I'm doing wrong, but these are the sorts of images I'm getting.
> ![image3](https://user-images.githubusercontent.com/61517961/266420779-329a84dc-f6b5-4833-929d-96b1a9a3bae7.jpg)
> ![image0](https://user-images.githubusercontent.com/61517961/266420892-2571a88e-1395-466f-989a-0861be417f47.jpg)
> ![image1-2](https://user-images.githubusercontent.com/61517961/266420917-c9a42b50-e7a1-4da3-b66d-da5809abc596.jpg)

**Comment by ViddleShtix** [comment]
The images with the code I posted look like normal SD1.5 photos they just don't resemble the subject I am training. I've tried training up to 5000 steps and still don't see any difference. Another thing I've noticed is that the saved weights are 2.5MB as opposed to the desired 100kb. In @shileims repo the weights are saved at 200kb which seems a lot closer.

I will try digging in again when I have some time.

**Comment by irowberry** [comment]
@BradVidler Yeah, that's interesting, my goal is to have it work with Hugging Face models, but I am getting similar results to you using your inference code, it seems that it isn't picking up on the concepts. I'll have to see what is wrong with my inference code, not sure why it isn't working.

**Comment by ViddleShtix** [comment]
Yes exactly. Making it work with HuggingFace is definitely ideal. My inference code is using HuggingFace's diffusers library directly whereas your code is implementing the under the hood code, rather than using the diffusers library which takes care of all that automatically.

Have you tested your inference code on default SD1.5 without any finetuning? Wondering if you get normal results that way.

EDIT: I believe I had a mistake in my inference code. I was passing in the non finetuned clip model into the OpenClipEmbedWrapper. After 2000 steps of training it *kind of* seems like it's getting closer. Here is I think the correct code:

wrapped_clip_with_new_concept = OpenClipEmbedWrapper(
    perfusion_model.clip_model,
    tokenizer = pipe.tokenizer,
    superclass_string = superclass_string
)

Also wondering if there is a way to specify batch size in the training. If it's defaulting to a batch size of 1 we might need to train a lot longer.

EDIT: Found the batch size but cannot increase it without running out of memory, probably due to have a Diffusers pipeline in memory.

I modified the inference function in your training code like so:
```
def inference(model):
    pipe = StableDiffusionPipeline.from_pretrained("runwayml/stable-diffusion-v1-5", requires_safety_checker=False)
    generator = torch.Generator(device="cuda").manual_seed(12345)
    prompts = ["photo of a person"]
    superclass_string = 'person'
    save_load.save(model, 'me_concept.pt')  
    inference_model = PerfusionModel(pipe.unet, pipe.text_encoder, pipe.tokenizer, superclass_string).to(device).requires_grad_(False)
    save_load.load(inference_model, 'me_concept.pt')
    inference_model.eval()
    inference_model.to("cpu").requires_grad_(False)
    wrapped_clip_with_new_concept = OpenClipEmbedWrapper(
        inference_model.clip_model,
        tokenizer = pipe.tokenizer,
        superclass_string = superclass_string
    )
    
    text_enc, superclass_enc, mask, indices = wrapped_clip_with_new_concept(prompts)

    pipe.unet = inference_model.unet
    pipe.text_encoder = inference_model.clip_model
    pipe.to(device)
    
    image = pipe(prompts[0], num_inference_steps=30, guidance_scale=7.5,generator=generator,cross_attention_kwargs={"concept_indices": indices, "text_enc_with_superclass": superclass_enc}).images[0]
    
    return image
```

and I am outputting a sample every 250 steps. I ran training for 8000 steps, using the same seed, and the output changes but it never ends up looking like the subject.

**Comment by ViddleShtix** [comment]
Another theory: your code is using EmbeddingWrapper, which doesn't create a new nn.identity for the new concept, whereas OpenClipEmbedWrapper does. Might be something to look into.

**Comment by irowberry** [comment]
I've discovered something interesting. Using the pipeline, there are two keyword arguments `prompts` and `prompt_embeds`. When using the prompts (text) I get an image that is a dog (the superclass) but not the one I trained it on. When using `prompt_embeds` I get a garbage image out, (it often triggers the NSFW detection filter for some reason). 
Here's my inference code. I'm not sure what is causing the difference, but I'll keep exploring

Update: If you pass in text to the pipe, it never calls any of the EmbeddingWrapper layer code. 
```
def inference(prompts, pipe, model):
    
    embeds_with_new_concept, embeds_with_superclass, embed_mask, concept_indices = model.wrapped_embeds(prompts)
    enc_with_new_concept = model.clip_model.text_model.encoder(embeds_with_new_concept).last_hidden_state
    
    if embeds_with_superclass is not None:
        enc_with_superclass = model.clip_model.text_model.encoder(embeds_with_superclass).last_hidden_state
    else:
        enc_with_superclass = embeds_with_superclass

    pipe.unet = model.unet
    pipe.text_encoder = model.clip_model
    pipe.to(device)
    images = pipe(prompt_embeds=enc_with_new_concept,
                 num_inference_steps=30, 
                 guidance_scale=7.5, 
                 cross_attention_kwargs={"concept_indices": concept_indices, "text_enc_with_superclass": enc_with_superclass}).images
    return images
```

**Comment by ViddleShtix** [comment]
Not having any luck here. Have also been testing @shileims repo to no avail. In that one I noticed it was using an outdated version of perfusion so I updated that and then had to make some slight modifications to make it work again. Trained for an hour on my 4090 and results don't look like the subject. I have also tried increasing the learning rates x10 and still got the same result.

Not too sure where to go from here.

**Comment by irowberry** [comment]
Yeah, me neither. I think the cross attention modules are training well, because the model still generates coherent pictures of whatever the superclass is, when I pass in text through the regular encoder. When I pass in the encodings/embeddings from the `EmbeddingWrapper` module into the pipe is when I get images of just essentially noise. @lucidrains any thoughts?

**Comment by ViddleShtix** [comment]
I spoke to Kohaku and they said that perfusion is incredibly complex and might require patching a lot of stuff to get the inference to work properly. They also suggested looking into pivotal tuning/rank1 lora which is implemented in hcpdiff so I'm changing focus to that.

If that doesn't work out I may commission someone to write the perfusion trainer. 

Feel free to reach out to me on Discord if you'd like to colab more. User name is Viddleshtix.

**Comment by lucidrains** [comment]
> Yeah, me neither. I think the cross attention modules are training well, because the model still generates coherent pictures of whatever the superclass is, when I pass in text through the regular encoder. When I pass in the encodings/embeddings from the `EmbeddingWrapper` module into the pipe is when I get images of just essentially noise. @lucidrains any thoughts?

sorry, been busy with other work

you should reach out to the author and see if he's willing to do a code review. i'm sure in the interest of his paper, he would want to see it reproduced in the wild

**Comment by IndolentKheper** [comment]
So nobody's been able to get this working, huh? A shame, a lot of us were hoping this would be the best new training method.

**Comment by paulinprogress** [comment]
> So nobody's been able to get this working, huh? A shame, a lot of us were hoping this would be the best new training method.

Heyo, fairly inexperienced research intern here, I've been working on this for the past 1-2 weeks. My main starting point was [this repo](https://github.com/ChenDarYen/Key-Locked-Rank-One-Editing-for-Text-to-Image-Personalization) to get things going as quickly as possible (i'm not really capable enough to set up my own training code from scratch yet).

I managed to get the training going, but with "mixed results" to say the least, so I'd love to have this whole discussion active again too. ^^ Ran into quite a few issues and i dont have *the best* hardware available either, but yeah see if that repo i linked is of any help.

**Comment by irowberry** [comment]
@PaulToast you can check out my repo and see if you can get it working. I might start working on it again. I think there is something wrong with the TextEmbedding wrapper, as the fine-tuned model generates normal images when using a non-trained text encoder.

**Comment by IndolentKheper** [comment]
> > So nobody's been able to get this working, huh? A shame, a lot of us were hoping this would be the best new training method.
> 
> Heyo, fairly inexperienced research intern here, I've been working on this for the past 1-2 weeks. My main starting point was [this repo](https://github.com/ChenDarYen/Key-Locked-Rank-One-Editing-for-Text-to-Image-Personalization) to get things going as quickly as possible (i'm not really capable enough to set up my own training code from scratch yet).
> 
> I managed to get the training going, but with "mixed results" to say the least, so I'd love to have this whole discussion active again too. ^^ Ran into quite a few issues and i dont have _the best_ hardware available either, but yeah see if that repo i linked is of any help.

Yeah I did check out that repo a while back, some of the same folks like shilheims were trying it but nobody's seemed to get anything useful working, then updates stopped, much like this one. I'm kind of skeptical about both key-locking papers regarding how much this is actually teaching a unique concept, versus image/token referencing and iterating, a la IP-adapter, from which I've gotten results very similar to input images. Is key-locking even that much different than IP-adapter, at this point?
